// @aureooms/js-collections-deque/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.EmptyDeque = exports.SingleElementDeque = exports.BoundedDeque = exports.UnboundedDeque = exports.ArbitrarySizeDeque = exports.Deque = exports._deque = exports.deque = undefined;

var _deque2 = require('./deque');

var _deque3 = _interopRequireDefault(_deque2);

var _deque4 = require('./_deque');

var _deque5 = _interopRequireDefault(_deque4);

var _implementation = require('./implementation');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = _deque3.default;
exports.deque = _deque3.default;
exports._deque = _deque5.default;
exports.Deque = _implementation.Deque;
exports.ArbitrarySizeDeque = _implementation.ArbitrarySizeDeque;
exports.UnboundedDeque = _implementation.UnboundedDeque;
exports.BoundedDeque = _implementation.BoundedDeque;
exports.SingleElementDeque = _implementation.SingleElementDeque;
exports.EmptyDeque = _implementation.EmptyDeque;
