// @aureooms/js-collections-deque/lib/implementation/EmptyDeque.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.default = EmptyDeque;

var _jsError = require('@aureooms/js-error');

var _Deque = require('./Deque');

var _Deque2 = _interopRequireDefault(_Deque);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function EmptyDeque(iterable) {

	if (iterable !== null) this.extend(iterable);
}

EmptyDeque.prototype = new _Deque2.default();

EmptyDeque.prototype.len = function () {

	return 0;
};

EmptyDeque.prototype.capacity = function () {

	return 0;
};

EmptyDeque.prototype.values = function () {

	return { next: function next() {
			return { done: true };
		} };
};

EmptyDeque.prototype.append = function (x) {

	return this;
};

EmptyDeque.prototype.appendleft = function (x) {

	return this;
};

EmptyDeque.prototype.clear = function () {

	return this;
};

EmptyDeque.prototype.copy = function () {

	return new EmptyDeque(this);
};

EmptyDeque.prototype._where = function (i) {

	throw new _jsError.IndexError(i);
};

EmptyDeque.prototype.pop = EmptyDeque.prototype.popleft = function () {

	throw new _jsError.IndexError("pop / popleft");
};
