// @aureooms/js-collections-deque/lib/implementation/SingleElementDeque.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.default = SingleElementDeque;

var _jsError = require('@aureooms/js-error');

var _Deque = require('./Deque');

var _Deque2 = _interopRequireDefault(_Deque);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function SingleElementDeque(iterable) {

	this.empty = true;

	this.value = 0;

	if (iterable !== null) this.extend(iterable);
}

SingleElementDeque.prototype = new _Deque2.default();

SingleElementDeque.prototype.len = function () {

	return this.empty ? 0 : 1;
};

SingleElementDeque.prototype.capacity = function () {

	return 1;
};

SingleElementDeque.prototype.values = regeneratorRuntime.mark(function _callee() {
	return regeneratorRuntime.wrap(function _callee$(_context) {
		while (1) {
			switch (_context.prev = _context.next) {
				case 0:
					if (!this.empty) {
						_context.next = 2;
						break;
					}

					return _context.abrupt('return');

				case 2:
					_context.next = 4;
					return this.value;

				case 4:
				case 'end':
					return _context.stop();
			}
		}
	}, _callee, this);
});

SingleElementDeque.prototype.append = SingleElementDeque.prototype.appendleft = function (x) {

	this.empty = false;
	this.value = x;

	return this;
};

SingleElementDeque.prototype.clear = function () {

	this.empty = true;
	this.value = 0;

	return this;
};

SingleElementDeque.prototype.copy = function () {

	return new SingleElementDeque(this);
};

SingleElementDeque.prototype.pop = SingleElementDeque.prototype.popleft = function () {

	if (this.empty) throw new _jsError.IndexError("pop / popleft");

	var value = this.value;

	this.empty = true;
	this.value = 0;

	return value;
};

SingleElementDeque.prototype.get = function (i) {

	if (this.empty || i !== 0) throw new _jsError.IndexError(i);

	return this.value;
};

SingleElementDeque.prototype.set = function (i, value) {

	if (this.empty || i !== 0) throw new _jsError.IndexError(i);

	this.value = value;

	return this;
};
